controllers/vgmanager.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	lvmv1alpha1 "github.com/red-hat-storage/lvm-operator/api/v1alpha1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type vgManager struct{}
+
+var _ resourceManager = vgManager{}
+
+const (
+	VGManagerUnit = "vg-manager"
+)
+
+func (v vgManager) getName() string {
+	return VGManagerUnit
+}
+
+func (v vgManager) ensureCreated(r *LVMClusterReconciler, ctx context.Context, lvmCluster *lvmv1alpha1.LVMCluster) error {
+	unitLogger := r.Log.WithValues("resourceManager", v.getName())
+	// aggregate nodeSelector and tolerations from all deviceClasses
+	nodeSelector, tolerations := extractNodeSelectorAndTolerations(*lvmCluster)
+	// get desired daemonset spec
+	dsTemplate := newVGManagerDaemonset(*lvmCluster)
+	// create desired daemonset or update mutable fields on existing one
+	ds := &appsv1.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      dsTemplate.Name,
+			Namespace: dsTemplate.Namespace,
+		},
+	}
+	unitLogger.Info("running CreateOrUpdate")
+	// the anonymous mutate function modifies the daemonset object after fetching it.
+	// if the daemonset does not already exist, it creates it, otherwise, it updates it
+	result, err := ctrl.CreateOrUpdate(ctx, r.Client, ds, func() error {
+		// at creation, deep copy the whole daemonset
+		if ds.CreationTimestamp.IsZero() {
+			dsTemplate.DeepCopyInto(ds)
```

Comment:
The newVGManagerDaemonset" does not include tolerations or node affinity, so .. if these terms are included in the topolvm cluster CRD from the very beginning, they are not going to be included in the daemonset in the creation, they will be included in a second pass (update) of the reconcile loop.

Reply:
Good catch.  I'm updated the initial deamonset to include toleration, nodeselectors, service account and the owner references.